app/controllers/admin/admin_controller.rb

```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+module Admin
+  class AdminController < ApplicationController
+    before_action :authenticate_admin_user!
+    layout 'admin'
+    def authenticate_admin_user!
+      current_user.present? ? redirect_to(root_path, alert: 'You are not authorized') : super
+    end
```

Comment:
Czy takie coś jest bezpieczne?

Reply:
Nie wiem za bardzo po co to jest. Tylko resource AdminUser powinien mieć tu dostęp. Powinienes to zalatwic chyba na poziomie devise w routsach

Reply:
@oskarlakner 
Dodałem tą linię, żeby spełnić to kryterium: 

> when a user tries to access /admin he should be redirected to / with a flash message "You are not authorized"

Więc w routsach dodałem 
```
namespace 'admin' do 
...  
end
```
Wszystko w `module Admin` dziedziczy z `AdminController` a tam jest `before_action :authenticate_admin_user!` i to zapewnia dostęp do `/admin` tylko dla AdminUser. 
Ale domyślnie `:authenticate_admin_user!` przekierowuje do `new_admin_user_session_path`, a wymóg mówi, że jeśli jest zalogowany zwykły `User` to przekierować go do `/` z flashem  `'You are not authorized'`. 
Jeśli jednak nikt nie jest zalogowany, to po wejściu na `/admin` przekierowany będzie do logowania na`new_admin_user_session_path`
Więc w skrócie po wejściu na `/admin`, jeśli: 

1. Nikt nie jest zalogowany - przekierowanie do logogwania dla admina
2. AdminUser jest zalogowany - dostęp jest zapewniony
3. zwykły User jest zalogowany - przekierwanie do `/`  z flashem 'You are not authorized'

Nie wiem czy ten wymóg dobrze zinterpretowałem.

Reply:
Nie zauważyłem, że dodałeś devise_for, jest ok 👍